### ISSUE TITLE
Investigate zfs as better alternative to RAID

### ISSUE THREAD

**Issue by alex-pakalniskis** [IndexerDAO/IndexerDAO#40]
## Resources
* https://wiki.ubuntu.com/ZFS

**Comment by alex-pakalniskis** [comment]
Low-ish priority but something we may want to consider for a third machine running Polygon

**Comment by alex-pakalniskis** [comment]
OpenZFS install docs for root on Ubuntu: https://openzfs.github.io/openzfs-docs/Getting%20Started/Debian/Debian%20Bullseye%20Root%20on%20ZFS.html

**Comment by alex-pakalniskis** [comment]
# [ZFS versus RAID: Eight Ironwolf disks, two filesystems, one winner](https://arstechnica.com/gadgets/2020/05/zfs-versus-raid-eight-ironwolf-disks-two-filesystems-one-winner/)

## RAID
A conventional RAID array is a simple abstraction layer that sits between a filesystem and a set of disks. It presents the entire array as a virtual "disk" device that, from the filesystem's perspective, is indistinguishable from an actual, individual disk—even if it's significantly larger than the largest single disk might be.

## ZFS
ZFS is an entirely different animal, and it encompasses functions that normally might occupy three separate layers in a traditional Unixlike system. It's a logical volume manager, a RAID system, and a filesystem all wrapped into one. Merging traditional layers like this has caused many a senior admin to grind their teeth in outrage, but there are very good reasons for it.

## Conclusions
If you're looking for raw, unbridled performance it's hard to argue against a properly-tuned pool of ZFS mirrors. RAID10 is the fastest per-disk conventional RAID topology in all metrics, and ZFS mirrors beat it resoundingly—sometimes by an order of magnitude—in every category tested, with the sole exception of 4KiB uncached reads.

**Comment by alex-pakalniskis** [comment]
# [RAID vs LVM vs ZFS Comparison](https://computingforgeeks.com/raid-vs-lvm-vs-zfs-comparison/)

## RAID
* It was basically developed to allow one to combine many inexpensive and small disks into an array in order to realize redundancy goals.

| RAID # | Redundancy Technique |
| --- | --- |
| RAID 0 | disk striping |
| RAID 1 | disk mirroring |
| RAID 5 | disk striping with parity |

## LVM
* abstraction of the idea of individual disk drives and allows you as the administrator to carve out “pieces” of space to use as drives. 
* It allows you to plug as many physical drives onto your individual system and then flexibly increase and decrease your logical volumes on your live host. 

## ZFS
* ZFS combines the roles of a file system and volume manager, enabling additional storage devices to be added to a live system and having the new space available on all of the existing file systems in that pool immediately. 
  * It does what LVM and RAID do in one package.
* with ZFS, the file system is aware of the underlying disk structure. This awareness makes the automatic growth of the existing file system possible for existing when additional disks are added to the pool.
* a number of different properties can be applied to each file system, hence the ability to create a number of different file systems and datasets rather than a single monolithic file system.
* ZFS implements RAID-Z, a variation on standard RAID-5 that offers better distribution of parity and eliminates the “RAID-5 write hole” in which the data and parity information become inconsistent in case of power loss.

**Comment by alex-pakalniskis** [comment]
# [An Introduction to ZFS A Place to Start](https://www.servethehome.com/an-introduction-to-zfs-a-place-to-start/)
## A filesystem and the logical volume manager
* ZFS directly controls not only how the bits and blocks of your files are stored on your hard drives, but it also controls how your hard drives are logically arranged for the purposes of RAID and redundancy. 
* ZFS can do some cool things like snapshots that a normal filesystem like NTFS could not. 
  * ZFS is classified as a copy-on-write or [COW filesystem](https://www.ixsystems.com/documentation/freenas/11.2/zfsprimer.html).

## Helpful terms
* A VDEV, or virtual device, is a logical grouping of one or more storage devices.
* A pool is then a logically defined group built from 1 or more VDEVs. 
  * ZFS is very customizable, and therefore, there are many different types of configurations for VDEVs.
* RAID-Z is a way of putting multiple drives together into a VDEV and storing parity, or fault tolerance. 

* What can make RAID-Z a better approach than a mirrored configuration is that it does not matter what drive fails in a RAID-Z. 
  * This benefit of RAID-Z comes at the cost of performance, however, and a mirrored pool will almost always be faster than RAID Z.

* RAID-Z is similar to a traditional RAID 5. In RAID-Z you have one drive worth of parity. 
* For RAID-Z you need a minimum of 3 drives per VDEV. You can have 3, 7, or even 12 drives in a RAID-Z VDEV.

**Comment by alex-pakalniskis** [comment]
# Notes from Launchpad Office Hours: October 19, 2022

They all keep talking about resilvering speed

## Matthew Darwin | Pinax Network
* zfs tip: don't zfs the root partition, just do the data disks.
* the man pages (docs) from zfs are pretty good
* It's a matter of finding the right settings for the specific workloads. (block size, compression, etc)
* create one data volume for each separate thing.
* I our zfs setup, we don't use disk mirroring.  Instead we run the same thing (eg an archive node) on multiple servers.
* So the redundancy is across servers, not across disks
* we use zfs striping across mutliple disks of course
```
zpool status
  pool: tank-ssd
 state: ONLINE
  scan: scrub repaired 0B in 07:16:04 with 0 errors on Sun Oct  9 07:40:06 2022
remove: Removal of vdev 3 copied 6.75T in 0h25m, completed on Fri Apr 29 01:13:37 2022
    2.30M memory used for removed device mappings
config:
```

* so here if one disk of the 4 here fails, then the whole thing is dead
* yep, raid 0
* redundancy across servers.
* blockchain data can always be re-created, so that's why we did it this way.

## GraphOps 
* uses RAIDZ pools
* Chris from GraphOps is looking at dRAID: https://openzfs.github.io/openzfs-docs/Basic%20Concepts/dRAID%20Howto.html
  * In summary dRAID can provide the same level of redundancy and performance as raidz, while also providing a fast integrated distributed spare.

### ZFS resources from Ana from GraphOps
* https://vadosware.io/post/everything-ive-seen-on-optimizing-postgres-on-zfs-on-linux/
* https://www.truenas.com/blog/zfs-pool-performance-2/
* https://www.truenas.com/blog/zfs-pool-performance-1/
* https://arstechnica.com/information-technology/2020/05/zfs-101-understanding-zfs-storage-and-performance/

## Jim | wavefive 
* is looking at GRAID as an alternative to RAIDZ
